force-app/main/default/classes/DonationHistoryController_TEST.cls

```diff
@@ -0,0 +1,120 @@
+
+        Contact mockDefaultContact = new Contact(
+            FirstName='Alice', 
+            LastName='Patterson', 
+            AccountId=mockDefaultAccount.Id
+        );
+        Contact secondaryContact = new Contact(
+            FirstName='Josh', 
+            LastName='Patterson',
+            AccountId=mockDefaultAccount.Id
+        );
+        insert mockDefaultContact;
+        insert secondaryContact;
+
+        Opportunity opportunityHardCredit = new Opportunity(
+            Name='HardCredit1', 
+            AccountId=mockDefaultAccount.Id, 
+            Primary_Contact__c=mockDefaultContact.Id, 
+            Amount=400, 
+            StageName='Closed Won', 
+            CloseDate=myDateTime
+        );
+        Opportunity opportunitySoftCredit = new Opportunity(
+            Name='SoftCredit1', 
+            AccountId=mockDefaultAccount.Id, 
+            Primary_Contact__c=secondaryContact.Id, 
+            Amount=200, 
+            StageName='Closed Won', 
+            CloseDate=myDateTime
+        );
+        Opportunity opportunityPartialSoftCredit = new Opportunity(
+            Name='PartialSoftCredit1', 
+            AccountId=mockDefaultAccount.Id, 
+            Primary_Contact__c=secondaryContact.Id, 
+            Amount=150, 
+            StageName='Closed Won', 
+            CloseDate=myDateTime
+        );
+        insert opportunityHardCredit;
+        insert opportunitySoftCredit;
+        insert opportunityPartialSoftCredit;
+
+        OpportunityContactRole oppCntRoleSoftCredit = new OpportunityContactRole(
+            Role='Soft Credit', 
+            OpportunityId=opportunitySoftCredit.Id, 
+            ContactId=mockDefaultContact.Id, 
+            isPrimary=false
+        );
+        OpportunityContactRole oppCntRolePartialSoftCredit = new OpportunityContactRole(
+            Role='Soft Credit', 
+            OpportunityId=opportunityPartialSoftCredit.Id, 
+            ContactId=mockDefaultContact.Id, 
+            isPrimary=false
+        );
+        insert oppCntRoleSoftCredit;
+        insert oppCntRolePartialSoftCredit;
+
+        Partial_Soft_Credit__c partialSoftCredit = new Partial_Soft_Credit__c (
+            Amount__c=50,
+            Contact__c=mockDefaultContact.Id,
+            Opportunity__c=opportunityPartialSoftCredit.Id,
+            Contact_Role_ID__c=oppCntRolePartialSoftCredit.Id
+        );
+        insert partialSoftCredit;
+        
+        Test.startTest();
+        List<DonationHistoryController.DonationHistory> donationHistory = DonationHistoryController.getDonationHistory(mockDefaultContact.Id);
+        List<DonationHistoryController.DonationHistory> filteredDonationHistory = new List<DonationHistoryController.DonationHistory>();
+        Test.stopTest();
+
+        for (DonationHistoryController.DonationHistory donation : donationHistory) {
+            filteredDonationHistory.add(donation);
+        }
+        System.assertEquals(3, donationHistory.size());
```

Comment:
I'm not sure I understand the objective here. The logic takes the result in the `donationHistory` collection and adds it to the `filteredDonationHistory` collection (in a loop that doesn't add any filtering or conditions) and then checks the size of the resulting `filteredDonationHistory`. Unit Tests need to be more explicit - testing very specific behaviors. Many unit tests won't even require DML because they can be passed mocked data to work with. In this case, I'm not sure this is actually testing anything at all other than that there are 3 records - and the test data setup is only setting up 3 records. 

There's no negative test methods either. For example, are there Opportunities or types of Soft Credits that should not be retrieved by the `getDonationHistory()` method? If so, then a new test method is needed to verify that the logic is _only_ pulling in the expected data and not data that should be filtered out.

Reply:
Validations against the data returned by the method were missing, as well as more test data to validate that it is being filtered correctly, which was the initial idea for this test.
more data was added to corroborate the filtering and also more assertions for the returned data